crates/telio-wg/src/pq.rs

```diff
@@ -0,0 +1,298 @@
+        noise::TunnResult::Err(err) => {
+            return Err(format!("Failed to decapsulate PQ keys message: {err:?}").into())
+        }
+        noise::TunnResult::WriteToTunnelV4(buf, _) => {
+            let ip = Ipv4Packet::new(buf).ok_or(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "Invalid PQ keys IP packet received",
+            ))?;
+
+            let udp = UdpPacket::new(ip.payload()).ok_or(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "Invalid PQ keys UDP packet received",
+            ))?;
+
+            let mut data = io::Cursor::new(udp.payload());
+
+            let mut version = [0u8; 4];
+            data.read_exact(&mut version)?;
+            let version = u32::from_le_bytes(version);
+            if version != HANDSHAKE_VERSION {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "Server responded with invalid PQ handshake version",
+                )
+                .into());
+            }
+
+            let mut ciphertext_len = [0u8; 4];
+            data.read_exact(&mut ciphertext_len)?;
+            let ciphertext_len = u32::from_le_bytes(ciphertext_len);
+
+            if ciphertext_len != CIPHERTEXT_LEN {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "Server responded with invalid PQ handshake ciphertext lenght",
+                )
+                .into());
+            }
+
+            data.read_exact(&mut cipherbuf)?;
+
+            kyber768::Ciphertext::from_bytes(&cipherbuf)
+                .map_err(|err| format!("Invalid PQ ciphertext received: {err:?}"))?
+        }
+        _ => return Err("Unexpected WG tunnel output".into()),
+    };
+
+    let pq_shared = kyber768::decapsulate(&ciphertext, &pq_secret);
+
+    Ok(PqKeys {
+        pq_shared,
+        wg_secret,
+    })
+}
+
+async fn handshake(
+    endpoint: impl ToSocketAddrs,
+    secret: &telio_crypto::SecretKey,
+    peers_pubkey: &telio_crypto::PublicKey,
+) -> Result<(Box<noise::Tunn>, UdpSocket), Error> {
+    let sock = UdpSocket::bind((Ipv4Addr::UNSPECIFIED, 0)).await?;
+    sock.connect(endpoint).await?;
+
+    let tunn = noise::Tunn::new(
+        secret.into_bytes().into(),
+        peers_pubkey.0.into(),
+        None,
+        None,
+        0,
+        None,
+    )?;
+
+    let mut pkgbuf = [0u8; 2048];
+    match tunn.encapsulate(&[], &mut pkgbuf) {
+        noise::TunnResult::Err(err) => {
+            return Err(format!("Failed to encapsulate handshake message: {err:?}").into())
+        }
+        noise::TunnResult::WriteToNetwork(buf) => {
+            sock.send(buf).await?;
+        }
+        _ => return Err("Unexpected WG tunnel output".into()),
+    }
+
+    // The response should be 92, so the buffer is sufficient
+    let read = sock.recv(&mut pkgbuf).await?;
```

Comment:
should this be wrapped in timeout?

Reply:
Yeah, I was thinking about it. However, I decided to wrap the whole method in the timeout on the higher level